rs/boundary_node/salt_sharing/canister/canister.rs

```diff
@@ -0,0 +1,161 @@
+
+// A small delay after beginning of the month when salt is generated
+const START_OF_MONTH_DELAY: Duration = Duration::from_secs(5);
+// Maximum number of attempts to initialize random seed
+const MAX_SEED_INIT_ATTEMPTS: i32 = 15;
+// Interval between retries to initialize random seed
+const SEED_INIT_RETRY_INTERVAL: Duration = Duration::from_secs(10);
+
+thread_local! {
+    // Random seed used for salt generation
+    static RNG: RefCell<Option<ChaCha20Rng>> = const { RefCell::new(None) };
+}
+
+// Run when the canister is first installed
+#[init]
+fn init(init_arg: InitArg) {
+    // Generate salt on the very first init or based on the provided argument.
+    if !is_salt_init() || init_arg.regenerate_now {
+        seed_rng_and_regenerate_salt(MAX_SEED_INIT_ATTEMPTS, Duration::ZERO);
+    }
+    // Start salt generation schedule based on the argument.
+    if let Some(strategy) = init_arg.salt_generation_strategy {
+        schedule_salt_generation(strategy);
+    }
+    // Start schedule of seeding randomness, if seeding_interval is None, schedule is executed only once.
+    let seeding_interval = init_arg.seeding_interval_secs.map(Duration::from_secs);
+    schedule_rng_seeding(Duration::ZERO, seeding_interval);
+}
+
+// Run every time a canister is upgraded
+#[post_upgrade]
+fn post_upgrade(init_arg: InitArg) {
+    // Run the same initialization logic
+    init(init_arg);
+}
+
+#[query]
+fn get_salt() -> GetSaltResponse {
+    get_salt_response()
+}
+
+// Fill RNG with pseudorandom bytes generated by the IC
+async fn seed_randomness() -> Result<(), String> {
+    let (seed,): ([u8; 32],) = ic_cdk::call(Principal::management_canister(), "raw_rand", ())
+        .await
+        .map_err(|err| {
+            format!(
+                "call to raw_rand of management canister failed: rejection_code={:?}, message={:?}",
+                err.0, err.1
+            )
+        })?;
+    RNG.with_borrow_mut(|rng| *rng = Some(ChaCha20Rng::from_seed(seed)));
+    Ok(())
+}
+
+// Schedules the seeding of randomness with an optional recurring period.
+fn schedule_rng_seeding(delay: Duration, period: Option<Duration>) {
+    set_timer(delay, move || {
+        spawn(async move {
+            if let Err(err) = seed_randomness().await {
+                // TODO: substitute with logs
+                ic_cdk::println!("Failed to seed randomness {err}");
+            }
+            // Schedule next seeding
+            if let Some(period) = period {
+                schedule_rng_seeding(period, Some(period));
+            }
+        })
+    });
+}
+
+// Sets an execution timer (delayed future task) and returns immediately.
+fn schedule_salt_generation(strategy: SaltGenerationStrategy) {
+    match strategy {
+        SaltGenerationStrategy::StartOfMonth => {
+            // Add a small delay to ensure execution starts after the month started
+            let delay = delay_till_next_month(time()) + START_OF_MONTH_DELAY;
```

Comment:
What are the requirements on START_OF_MONTH_DELAY? Is it supposed to prevent a clock drift of the subnet?

Reply:
The idea is to avoid multiple `regenerate_salt()` invocations at the end of the same month due to numerical errors and potentially non-monotonically increasing time.

Reply:
> numerical errors

what do you mean? bugs in the function computing the end of the month?

> potentially non-monotonically increasing time

this would be a critical error in the replica; so it should be safe to assume that time is (strictly) monotone